### ISSUE TITLE
out of memory

### ISSUE THREAD

**Issue by MarioProjects** [HaoMood/bilinear-cnn#13]
Hi! After 2 epochs the backward runs out of memory :( First epoch its okey but then crash on second one. It seems that stores the graph or somethin but I change some things and crash:

`
    for X, y in self._train_loader:
                # Data.
                
                # Clear the existing gradients.
                X = X.cuda()
                y = y.cuda()

                # Forward pass.
                score = self._net(X)
                loss = self._criterion(score, y.long())

                with torch.no_grad():
                    epoch_loss += loss.item()
                    # Prediction.
                    prediction = torch.argmax(score, dim=1)
                    num_total += y.size(0)
                    num_correct += torch.sum(prediction == y.long()).item()

                # Backward pass.
                self._optimizer.zero_grad()
                loss.backward()
                self._optimizer.step()
                
                total_batches+=1
                del X, y, score, loss, prediction
`

**Comment by HaoMood** [comment]
Your suspicion is reasonable. It may accumulate history across the training loop.

However, I have not found the bug in the code neither. Loss and num_correct are all use .item() function, which sounds good.

**Comment by ewyuanzhang** [comment]
The memory problem may come from the testing phase. Per discussion [here](https://discuss.pytorch.org/t/model-eval-vs-with-torch-no-grad/19615/2), PyTorch still computes the gradients of model parameters with `model.train(False)`. During training, the gradients will be freed at `loss.backward()`, but since loss is not computed in testing, gradients stay in the memory forever. Wrapping the codes with `torch.no_grad()` stops PyTorch from computing the gradients and hence solves the problem.
```python
model.train(False)
with torch.no_grad():
    for X, y in test_loader:
        ...
model.train(True)
```

**Comment by HaoMood** [comment]
Thanks for your comment. 

This code is developmented under PyTorch 0.3.0 (you can infer that from the usage like X = torch.autograd.Variable(X)). However, the torch.no_grad() context is available in PyTorch version >= 0.4.0. In PyTorch version >= 0.4.0 it is very neccessary to enclose the testing phase under torch.no_grad().

You may refer to this repo https://github.com/HaoMood/blinear-cnn-faster for running your code under PyTorch version >= 0.4.0, and this repo also runs faster.
